### PR TITLE
Add gaswater solvent

### PR DESCRIPTION
**Pull request by totto82**
Fix some places that assumed 3p + solvent

See https://github.com/OPM/opm-simulators/issues/4548

Depends on https://github.com/OPM/opm-models/pull/796

Compiles but does not converge.

### CODE REVIEW

**Comment by totto82** [comment]
Gives same results as a test case with trivial oil phase. i.e. this is ready for review.

**Comment by totto82** [comment]
jenkins build this please

File: opm/simulators/wells/SingleWellState.cpp [hunk 1]
```diff
@@ -161,6 +161,7 @@ double SingleWellState::sum_polymer_rates() const {
 
 
 double SingleWellState::sum_solvent_rates() const {
+    //return this->surface_rates[pu.phase_pos[BlackoilPhases::Vapour]];
```
Comment (GitPaean) [inline opm/simulators/wells/SingleWellState.cpp:164]: please remove this commented line.

File: opm/simulators/wells/StandardWellPrimaryVariables.cpp [hunk 2]
```diff
@@ -455,6 +461,9 @@ volumeFraction(const unsigned compIdx) const
         if (has_gfrac_variable && compIdx == Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx)) {
             return evaluation_[GFrac];
         }
+        if (Indices::enableSolvent && compIdx == (unsigned)Indices::contiSolventEqIdx) {
+            return evaluation_[SFrac];
+        }
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWellPrimaryVariables.cpp:466]: can we reduce the duplication of the following code? 

```c++
        if (has_gfrac_variable && compIdx == Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx)) {
            return evaluation_[GFrac];
        }
        if (Indices::enableSolvent && compIdx == (unsigned)Indices::contiSolventEqIdx) {
            return evaluation_[SFrac];
        }
```

File: opm/simulators/wells/StandardWellPrimaryVariables.cpp [hunk 3]
```diff
@@ -475,6 +484,10 @@ volumeFraction(const unsigned compIdx) const
     else if (FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx) &&
              FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
         well_fraction -= evaluation_[GFrac];
+
+        if constexpr (Indices::enableSolvent) {
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWellPrimaryVariables.cpp:488]: I think we also only need one the following code? 

```c++
        if constexpr (Indices::enableSolvent) {
            well_fraction -= evaluation_[SFrac];
        }
```

Maybe it might cause problem for single phase solvent case?

**Review by GitPaean — CHANGES_REQUESTED** [review]
The code looks fine, but in general with all the phase combinations, the code in general (not only in this PR) is very hard to follow and easy to make mistakes now.  Maybe we should clear up the logic so it can be clearer to follow.

**Comment by totto82** [comment]
Thanks for the review. I have followed your suggestions.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
I plan to add the test case in  https://github.com/OPM/opm-tests/pull/973 when it gets merged.

File: opm/simulators/wells/StandardWellPrimaryVariables.cpp [hunk 4]
```diff
@@ -438,45 +444,31 @@ volumeFraction(const unsigned compIdx) const
...
-        }
-
-        if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
-            well_fraction -= evaluation_[GFrac];
-        }
-
-        if constexpr (Indices::enableSolvent) {
-            well_fraction -= evaluation_[SFrac];
-        }
+    if (has_wfrac_variable && FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx)) {
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWellPrimaryVariables.cpp:463]: can you modify the comment `// Oil or WATER fraction` to be more specific. 

Basically, it will be oil fraction when oil phase is active and water fraction when oil phase is not active.

**Comment by GitPaean** [comment]
Thanks for the update. The PR looks fine to me now except the small request regarding comment (some comment is needed to make the code more maintainable). 

Then we should wait the testing case going in the opm-tests and create the regression test to protect the development.

**Comment by totto82** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
https://github.com/OPM/opm-tests/pull/973 just got merged. Please add the regressing test, register the reference update, so we can update the reference and get this PR in.

**Review by GitPaean — APPROVED** [review]

**Comment by totto82** [comment]
jenkins build this opm-tests=979 please
